graphql_compiler/query_formatting/common.py

```diff
@@ -1,18 +1,106 @@
+    elif GraphQLFloat.is_same_type(stripped_type):
+        if not isinstance(value, float):
+            raise GraphQLInvalidArgumentError(u'Attempting to represent a non-float as a float: '
+                                              u'{} {}'.format(type(value), value))
+    elif GraphQLInt.is_same_type(stripped_type):
+        # Special case: in Python, isinstance(True, int) returns True.
+        # Safeguard against this with an explicit check against bool type.
+        if isinstance(value, bool) or not isinstance(value, six.integer_types):
+            raise GraphQLInvalidArgumentError(u'Attempting to represent a non-int as an int: '
+                                              u'{} {}'.format(type(value), value))
+    elif GraphQLBoolean.is_same_type(stripped_type):
+        if not isinstance(value, bool):
+            raise GraphQLInvalidArgumentError(u'Attempting to represent a non-bool as a bool: '
+                                              u'{} {}'.format(type(value), value))
+    elif GraphQLDecimal.is_same_type(stripped_type):
+        # Types we support are int, float, and Decimal, but not bool.
+        # isinstance(True, int) returns True, so we explicitly forbid bool.
+        if isinstance(value, bool):
+            raise GraphQLInvalidArgumentError(
+                u'Attempting to represent a non-decimal as a decimal: {} {}'
+                .format(type(value), value))
+        if not isinstance(value, decimal.Decimal):
+            try:
+                decimal.Decimal(value)
+            except decimal.InvalidOperation as e:
+                raise GraphQLInvalidArgumentError(e)
+    elif GraphQLDate.is_same_type(stripped_type):
+        # Datetimes pass as instances of date. We want to explicitly only allow dates.
+        if isinstance(value, datetime.datetime) or not isinstance(value, datetime.date):
+            raise GraphQLInvalidArgumentError(u'Attempting to represent a non-date as a date: '
+                                              u'{} {}'.format(type(value), value))
+        try:
+            stripped_type.serialize(value)
+        except ValueError as e:
+            raise GraphQLInvalidArgumentError(e)
+    elif GraphQLDateTime.is_same_type(stripped_type):
+        if not isinstance(value, (datetime.date, arrow.Arrow)):
+            raise GraphQLInvalidArgumentError(
+                u'Attempting to represent a non-datetime as a datetime: {} {}'
+                .format(type(value), value))
+        try:
+            stripped_type.serialize(value)
+        except ValueError as e:
+            raise GraphQLInvalidArgumentError(e)
+    elif isinstance(stripped_type, GraphQLList):
+        if not isinstance(value, list):
+            raise GraphQLInvalidArgumentError(u'Attempting to represent a non-list as a list: '
+                                              u'{} {}'.format(type(value), value))
+        inner_type = strip_non_null_from_type(stripped_type.of_type)
```

Comment:
Can you delete the `strip_non_null_from_type` here? There is a `strip_non_null_from_type` at the beginning of the function anyways so this doesn't do anything.

Reply:
It definitely does do something :)

`GraphQLNonNull(GraphQLList(GraphQLNonNull(GraphQLString)))` is a valid type. The first `strip_non_null_from_type` will cause `stripped_type = GraphQLList(GraphQLNonNull(GraphQLString))`. This second call takes `stripped_type.of_type` and strips that, to cause `inner_type = GraphQLString`. If I remove this call, we'll have `inner_type = GraphQLNonNull(GraphQLString)` which will then fail the new test I added.

Reply:
No it will not. We will pass `GraphQLNonNull(GraphQLString)` into `validate_argument_type` which will then create the variable `stripped_type` with value `GraphQLString` and validate the argument based on the `stripped_type`.

Reply:
I agree, it won't fail, so I take that part back. But why unwrap the type on every single iteration, when we can unwrap it once and for all? That loop is `O(len(value))` and `value` is user-supplied so it may be quite large, and it doesn't cost us much to avoid a bunch of inner loop work.